src/Strategy/AlpineLinux/Parser.hs

```diff
@@ -0,0 +1,128 @@
+-- | Parsing functions for APK files.
+-- APK is the package manager for Alpine Linux (https://alpinelinux.org/).
+--
+-- The database itself is just a flat file with entries in text.  These are
+-- specified in a key-value format with obscure names.  See
+-- https://wiki.alpinelinux.org/wiki/Apk_spec#Syntax.
+module Strategy.AlpineLinux.Parser (installedPackagesDatabaseParser, PackageError (..)) where
+
+import Data.List.NonEmpty (NonEmpty (..))
+import Data.List.NonEmpty qualified as NE
+import Data.Map qualified as Map
+import Data.String.Conversion (toText)
+import Data.Text (Text)
+import Data.Void (Void)
+import Strategy.AlpineLinux.Types (AlpinePackage (..))
+import Text.Megaparsec (
+  MonadParsec (eof, takeWhileP),
+  Parsec,
+  many,
+  optional,
+  some,
+  try,
+ )
+import Text.Megaparsec.Char (char, eol, letterChar)
+
+-- | Errors that can occur even in a valid file.  Packages are a lose collection
+-- of fields, but it unclear which are required.  These error represent fields
+-- we expect, but which may be missing in unusual circumstances.
+data PackageError
+  = MissingPackageName
+  | MissingPackageArchitecture {packageName :: Text}
+  | MissingPackageVersion {packageName :: Text}
+  deriving (Show, Eq, Ord)
+
+type Parser = Parsec Void Text
+
+-- | Key for the package-name field.
+packageNameKey :: Text
+packageNameKey = "P"
+
+-- | Key for the architecture field.
+architectureKey :: Text
+architectureKey = "A"
+
+-- | Key for the version field.
+versionKey :: Text
+versionKey = "V"
+
+-- | Parses zero or more occurrences of a construct with separators between.
+--
+-- This differs from `Control.Monad.Combinators.sepBy` because it will backtrack
+-- failed repetitions.  The generic version has no notion of backtracking so if
+-- the separator consumes input after the last entry then the whole parser will
+-- fail because it will fail to parse another entry after the separator.
+--
+-- TODO: If this is more generally useful we should move it to a common location.
```

Comment:
I think the inner usages of `try` would make me want to avoid using this generally.

Reply:
It makes the semantic similar to what we get in other parsers like attoparsec, which might be more intuitive.  I find it makes more sense to work this way.

One weird thing is that different constructs have different backtracking behavior in Megaparsec.  (From the docs) "As of Megaparsec 4.4.0, string backtracks automatically (see tokens), so it does not need try."  This means that you might refactor something slightly to have a compound parsers instead of a primitive one and suddenly your `sepBy` would break in surprising ways.  I think this version is "safer" and less error prone.

Also, being restricted to Megaparsec with that `try` at least means it's not going to cause changes to other monads.